nova/virt/vmwareapi/driver.py

```diff
@@ -690,6 +693,44 @@ def detach_interface(self, context, instance, vif):
         """Detach an interface from the instance."""
         self._vmops.detach_interface(context, instance, vif)
 
+    def sync_server_group(self, context, sg_uuid):
+        self._vmops.sync_server_group(context, sg_uuid)
+
+    def _server_group_sync_loop(self, context):
+        """Retrieve all groups from the cluster and from the DB and call
+        self.sync_server_group() for them.
+
+        We always wait a little not not overwhelm the cluster.
+        """
+        InstanceList = objects.instance.InstanceList
+        instance_uuids = set(InstanceList.get_by_host(
+            context, self._compute_host, expected_attrs=[]))
+
+        InstanceGroupList = objects.instance_group.InstanceGroupList
+        ig_list = InstanceGroupList.get_by_isntance_uuids(
+            context, instance_uuids)
+
+        sg_uuids = set(ig.uuid for ig in ig_list)
+
+        cluster_rules = cluster_util.fetch_cluster_rules(
+            self._session, cluster_ref=self._cluster_ref)
+        for rule in cluster_rules:
+            if not rule.name.startswith(constants.DRS_PREFIX):
+                # not managed by us
+                continue
+
+            # should look like
+            # NOVA_d7134f26-f2e2-42ed-b5f1-4092962e84d5-affinity
+            parsed_uuid = rule.name[len(constants.DRS_PREFIX):]
+            # and now d7134f26-f2e2-42ed-b5f1-4092962e84d5-affinity
+            parsed_uuid = '-'.join(parsed_uuid.split('-')[:5])
```

Comment:
Agreed, that this might make sense with the given data. I'll add another example to the comment (we also have `-soft-anti-affinity` and `-anti-affinity` so the last part is not always the same.)

Reply:
Wouldn't it then make sense to simply have a regex capture the uuid?

Reply:
Might. I'll test what's faster.

Reply:
```
In [13]: %timeit re.search(UUID_RE_STR, a).group(0)                                                                                          
1.12 µs ± 57.3 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
In [18]: %timeit '-'.join(a[len(DRS_PREFIX):].split('-')[:5])                                                                                
463 ns ± 14.8 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```
from this perspective, it doesn't make sense.

Reply:
If you compile the regex, it looks more like
```
In [2]: UUID_RE = re.compile(r'[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}')
In [6]: timeit UUID_RE.search(a).group(0)
571 ns ± 3.27 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each
In [8]: timeit '-'.join(a[len(DRS_PREFIX):].split('-')[:5])
551 ns ± 3.03 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

But my main reason is not so much speed. I find a regex a bit more readable, in that sense that the code directly will say that you look for a UUID in the name. The regex itself is of course often not as easy to parse "manually".

Reply:
Interesting, I though Python would cache used regexes and thus pasted the second try instead of the first.